imageutils.py

```diff
@@ -845,25 +854,38 @@ def generate_splits_name(y_size, z_size, x_size, Y_size, Z_size, X_size, out_dir
     return split_names
 
 
-def generate_legend_file(split_names, legend_file_name, out_dir):
+def generate_legend_file(split_names, legend_file_name, out_dir, hdfs_client=None):
     """
     generate legend file for each all the splits
     """
     legend_file = '{0}/{1}'.format(out_dir, legend_file_name)
-    with open(legend_file, 'a+') as f:
-        for split_name in split_names:
-            f.write('{0}\n'.format(split_name))
-    return legend_file
 
+    if not hdfs_client:
+        with open(legend_file, 'a+') as f:
+            for split_name in split_names:
+                f.write('{0}\n'.format(split_name))
+    else:
+        with hdfs_client.write(legend_file) as f:
+            for split_name in split_names:
+                f.write('{0}\n'.format(split_name))
+
+    return legend_file
 
-def generate_headers_of_splits(split_names, y_size, z_size, x_size, dtype):
+def generate_headers_of_splits(split_names, y_size, z_size, x_size, dtype, hdfs_client=None):
     """
     generate headers of each splits based on the shape and dtype
     """
     header = generate_header(y_size, z_size, x_size, dtype)
-    for split_name in split_names:
-        with open(split_name, 'w+b') as f:
-            header.write_to(f)
+
+    if not hdfs_client:
```

Comment:
there's redundancy with the previous method here. You could write a single method.